File: 351004/Kadushko/Lab1/restApi/src/main/java/com/example/restApi/services/IssueService.java
package com.example.restApi.services;

import com.example.restApi.dto.request.IssueRequestTo;
import com.example.restApi.dto.response.IssueResponseTo;
import com.example.restApi.exception.AlreadyExistsException;
import com.example.restApi.exception.NotFoundException;
import com.example.restApi.mapper.IssueMapper;
import com.example.restApi.model.Issue;
import com.example.restApi.repository.EditorRepository;
import com.example.restApi.repository.IssueRepository;
import org.springframework.stereotype.Service;

import java.time.LocalDateTime;
import java.util.HashSet;
import java.util.List;
import java.util.stream.Collectors;

@Service
public class IssueService {

    private final IssueRepository issueRepository;
    private final EditorRepository editorRepository;
    private final IssueMapper issueMapper;

    public IssueService(IssueRepository issueRepository,
                        EditorRepository editorRepository,
                        IssueMapper issueMapper) {
        this.issueRepository = issueRepository;
        this.editorRepository = editorRepository;
        this.issueMapper = issueMapper;
    }

    public List<IssueResponseTo> getAll() {
        return issueRepository.findAll().stream()
                .map(issueMapper::toDto)
                .collect(Collectors.toList());
    }

    public IssueResponseTo getById(Long id) {
        Issue issue = issueRepository.findById(id)
                .orElseThrow(() -> new NotFoundException("Issue not found with id: " + id));
        return issueMapper.toDto(issue);
    }

    public IssueResponseTo create(IssueRequestTo request) {
        if (!editorRepository.existsById(request.getEditorId())) {
            throw new NotFoundException("Editor not found with id: " + request.getEditorId());
        }
        if (issueRepository.existsByTitle(request.getTitle())) {
            throw new AlreadyExistsException("Issue with title '" + request.getTitle() + "' already exists");
        }
        Issue issue = issueMapper.toEntity(request);
        issue.setId(null);
        issue.setCreated(LocalDateTime.now());
        issue.setModified(LocalDateTime.now());
        if (issue.getMarkerIds() == null) {
            issue.setMarkerIds(new HashSet<>());
        }
        Issue saved = issueRepository.save(issue);
        return issueMapper.toDto(saved);
    }

    public IssueResponseTo update(IssueRequestTo request) {
        if (request.getId() == null) {
            throw new IllegalArgumentException("Issue ID is required for update");
        }
        Issue issue = issueRepository.findById(request.getId())
                .orElseThrow(() -> new NotFoundException("Issue not found with id: " + request.getId()));
        if (!editorRepository.existsById(request.getEditorId())) {
            throw new NotFoundException("Editor not found with id: " + request.getEditorId());
        }
        if (issueRepository.existsByTitleAndIdNot(request.getTitle(), request.getId())) {
            throw new AlreadyExistsException("Issue with title '" + request.getTitle() + "' already exists");
        }
        LocalDateTime created = issue.getCreated();
        issueMapper.updateEntity(issue, request);
        issue.setCreated(created);
        issue.setModified(LocalDateTime.now());
        if (issue.getMarkerIds() == null) {
            issue.setMarkerIds(new HashSet<>());
        }
        Issue updated = issueRepository.save(issue);
        return issueMapper.toDto(updated);
    }

    public void delete(Long id) {
        if (!issueRepository.existsById(id)) {
            throw new NotFoundException("Issue not found with id: " + id);
        }
        issueRepository.deleteById(id);
    }
}

File: 351004/Barsukov/app/services/issue_service.py
from datetime import datetime

from app.repository.memory import issue_repo, author_repo
from app.schemas.issue import IssueRequestTo, IssueResponseTo
from app.core.exceptions import AppException


class IssueService:
    def create(self, dto: IssueRequestTo) -> IssueResponseTo:
        # Проверяем, что автор существует
        if not author_repo.find_by_id(dto.authorId):
            raise AppException(400, "Author not found", 4)

        now = datetime.now().isoformat()
        data = dto.model_dump()
        data.update({"created": now, "modified": now})

        return IssueResponseTo(**issue_repo.save(data))

    def get_all(self):
        return [IssueResponseTo(**i) for i in issue_repo.find_all()]

    def get_by_id(self, id: int):
        res = issue_repo.find_by_id(id)
        if not res:
            raise AppException(404, "Issue not found", 5)
        return IssueResponseTo(**res)

    def update(self, id: int, dto: IssueRequestTo) -> IssueResponseTo:
        # Автор должен существовать
        if not author_repo.find_by_id(dto.authorId):
            raise AppException(400, "Author not found", 7)

        existing = issue_repo.find_by_id(id)
        if not existing:
            raise AppException(404, "Issue not found", 6)

        now = datetime.now().isoformat()
        data = dto.model_dump()
        data["id"] = id
        data["created"] = existing.get("created", now)
        data["modified"] = now

        return IssueResponseTo(**issue_repo.save(data))

    def delete(self, id: int):
        if not issue_repo.delete(id):
            raise AppException(404, "Issue not found", 8)
